pghoard/walreceiver.py

```diff
@@ -0,0 +1,206 @@
+        assert len(rest_of_data) == 0
+        # Pad with 0s up to XLOG_SEG_SIZE
+        self.buffer = BytesIO(rest_of_data)
+        self.buffer.seek(0, 2)
+
+    def run(self):
+        self._get_cursor()
+        if self.replication_slot:
+            self.create_replication_slot()
+        timeline = self.start_replication()
+        while self.running:
+            wal_name = None
+            try:
+                msg = self.c.read_message()
+            except psycopg2.DatabaseError:
+                self.log.exception("Problem reading msg")
+                continue
+            self.log.debug("replication_msg: %r, buffer: %r/%r",
+                           msg, self.buffer.tell(), XLOG_SEG_SIZE)
+            if msg:
+                self.latest_activity = datetime.datetime.utcnow()
+                log, _, seg = convert_integer_to_lsn(msg.data_start)
+                wal_name = name_for_tli_log_seg(timeline, log, seg)
+
+                if not self.latest_wal:
+                    self.latest_wal_start = msg.data_start
+                    self.latest_wal = wal_name
+                self.buffer.write(msg.payload)
+
+                # TODO: Calculate end pos and transmit that?
+                msg.cursor.send_feedback(write_lsn=msg.data_start)
+
+            if wal_name and self.latest_wal != wal_name or self.buffer.tell() >= XLOG_SEG_SIZE:
+                self.switch_xlog()
+
+            for wal_start, queue in self.callbacks.items():
+                try:
+                    transfer_result = queue.get_nowait()
+                    self.log.debug("Transfer result: %r", transfer_result)
+                    self.completed_wal_segments.add(wal_start)
+                except Empty:
+                    continue
+
+            for completed_lsn in sorted(self.completed_wal_segments):
+                self.callbacks.pop(completed_lsn)
+                if self.callbacks:
+                    if completed_lsn > min(self.callbacks):
+                        pass  # Do nothing since a smaller lsn is still being transferred
+                    else:  # Earlier lsn than earlist on-going transfer, just advance flush_lsn
+                        self.c.send_feedback(flush_lsn=completed_lsn)
+                        self.completed_wal_segments.discard(completed_lsn)
+                        self.last_flushed_lsn = completed_lsn
+                        self.log.debug("Sent flush_lsn feedback as: %r", self.last_flushed_lsn)
+                else:  # No on-going transfer, just advance flush_lsn
+                    self.c.send_feedback(flush_lsn=completed_lsn)
+                    self.completed_wal_segments.discard(completed_lsn)
+                    self.last_flushed_lsn = completed_lsn
+                    self.log.debug("Sent flush_lsn feedback as: %r", self.last_flushed_lsn)
+
+            if not msg:
+                timeout = KEEPALIVE_INTERVAL - (datetime.datetime.now() - self.c.io_timestamp).total_seconds()
+                try:
```

Comment:
`with suppress(InterruptedError):` if you really just want to ignore it instead of restarting select.  considering python 3.5 restarts calls automatically on InterruptedError we should probably make a restarting_select function in compat.py and use it.